Quaternion/Quaternion.py

```diff
@@ -0,0 +1,513 @@
+
+        :rtype: numpy array
+
+        """
+        if self._q is None:
+            # Figure out q from available values, doing nothing others are not defined
+            if self._equatorial is not None:
+                self._q = self._equatorial2quat()
+            elif self._T is not None:
+                self._q = self._transform2quat()
+        return self._q
+
+    # use property to make this get/set automatic
+    q = property(_get_q, _set_q)
+
+    def __repr__(self):
+        q = self.q
+        return ('<{} q1={:.8f} q2={:.8f} q3={:.8f} q4={:.8f}>'
+                .format(self.__class__.__name__, q[0], q[1], q[2], q[3]))
+
+    def _set_equatorial(self, equatorial):
+        """Set the value of the 3 element equatorial coordinate list [RA,Dec,Roll]
+           expects values in degrees
+           bounds are not checked
+
+           :param equatorial: list or array [ RA, Dec, Roll] in degrees
+
+        """
+        att = np.array(equatorial)
+        ra, dec, roll = att
+        self._ra0 = ra
+        if (ra > 180):
+            self._ra0 = ra - 360
+            self._roll0 = roll
+        if (roll > 180):
+            self._roll0 = roll - 360
+        self._equatorial = att
+
+    def _get_equatorial(self):
+        """Retrieve [RA, Dec, Roll]
+
+        :rtype: numpy array
+        """
+        if self._equatorial is None:
+            if self._q is not None:
+                self._equatorial = self._quat2equatorial()
+            elif self._T is not None:
+                self._q = self._transform2quat()
+                self._equatorial = self._quat2equatorial()
+        return self._equatorial
+
+    equatorial = property(_get_equatorial, _set_equatorial)
+
+    def _get_ra(self):
+        """Retrieve RA term from equatorial system in degrees"""
+        return self.equatorial[0]
+
+    def _get_dec(self):
+        """Retrieve Dec term from equatorial system in degrees"""
+        return self.equatorial[1]
+
+    def _get_roll(self):
+        """Retrieve Roll term from equatorial system in degrees"""
+        return self.equatorial[2]
+
+    ra = property(_get_ra)
+    dec = property(_get_dec)
+    roll = property(_get_roll)
+
+    def _get_zero(self, attr):
+        """
+        Return a version of attr that is between -180 <= val < 180
+        """
+        if not hasattr(self, '_' + attr):
+            val = getattr(self, attr) % 360.0
+            if val >= 180:
+                val -= 360
+        return val
+
+    @property
+    def ra0(self):
+        """
+        Return quaternion RA in the range -180 <= roll < 180.
+        """
+        if not hasattr(self, '_ra0'):
+            self._ra0 = self._get_zero('ra')
+        return self._ra0
+
+    @property
+    def roll0(self):
+        """
+        Return quaternion roll in the range -180 <= roll < 180.
+        """
+        if not hasattr(self, '_roll0'):
+            self._roll0 = self._get_zero('ra')
```

Comment:
Looks like it passed the new ra0 roll0 test because ra0 and roll0 were equal in the test.